### PR TITLE
[sgl-kernel] support > 1024 experts in moe_align_block_size kernel

### PR DESCRIPTION
**Pull request by klshuster**
## Motivation

  The existing `moe_align_block_size` CUDA kernel uses CUB block-level scan primitives that are limited to 1024 threads, which caps the maximum number of experts at 1024. Models with virtual/merged LoRA experts or very large MoE configurations (e.g., 2048 or 4096 experts) hit this limit. This PR adds a v2 kernel path that supports up to 4096 experts.

  ## Modifications

  - **New `moe_align_block_size_kernel_v2` kernel** (`sgl-kernel/csrc/moe/moe_align_kernel.cu`): Uses an `EXPERTS_PER_THREAD` template parameter (2 or 4) so that 1024 threads can handle up to 4096 experts. Replaces CUB with a two-level warp exclusive prefix sum — intra-warp scan followed by a cross-warp scan in warp 0 — to avoid the 1024-element CUB limitation.
- **Dispatch logic**: The existing kernel handles `num_experts <= 1024`. The v2 kernel is dispatched for `num_experts > 1024` (with `EXPERTS_PER_THREAD=2` for ≤2048, `EXPERTS_PER_THREAD=4` for ≤4096). A `TORCH_CHECK` enforces the 4097 upper bound.
- **Moved `warp_exclusive_scan`** out of `#ifdef __CUDA_ARCH__` guards so it can be used by both the original and v2 kernels.
- **New test** (`sgl-kernel/tests/test_moe_align.py`): `test_moe_align_block_size_v2_large_num_experts` validates the v2 kernel against the Triton reference for expert counts of 1025, 2048, and 4096.

## Accuracy Tests

The new test (`test_moe_align_block_size_v2_large_num_experts`) compares CUDA v2 output against the Triton reference implementation across combinations of `block_size={64,128}`, `num_tokens={1,8,32,256}`, `topk=8`, `num_experts={1025,2048,4096}` — checking `num_tokens_post_pad`, `expert_ids`, and per-block `sorted_token_ids` for exact equality.

## Speed Tests and Profiling
N/A — this is a correctness extension to support larger expert counts; the kernel launch configuration (2 blocks, 1024 threads) is lightweight and not on the latency-critical path for typical workloads.
## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

### CODE REVIEW

**Review by sshleifer — APPROVED** [review]

**Review by BBuf — COMMENTED** [review]
`moe_align_block_size` is called with `num_experts + 1` to reserve bucket `0` for the filtered `-1` expert ID, so `4096` real experts means the kernel actually needs to handle `4097` internal buckets. However, the v2 kernel launches `1024` threads with `EXPERTS_PER_THREAD=4`, which covers at most `4096` bucket indices (`0..4095`). That leaves bucket `4096` uncovered, which corresponds to the last real expert after the `+1` offset convention. As a result, the prefix/cumsum for the last expert may never be written correctly, so the `4096`-expert case appears unsupported despite the new `TORCH_CHECK(num_experts <= 4097)`. Can you confirm this?

File: python/sglang/jit_kernel/tests/test_moe_align_block_size.py [hunk 1]
```diff
@@ -0,0 +1,345 @@
+import itertools
+import sys
+
+import pytest
+import torch
+import triton
+import triton.language as tl
+
+from sglang.jit_kernel.moe_align import moe_align_block_size
+
```
Comment (yushengsu-thu) [inline python/sglang/jit_kernel/tests/test_moe_align_block_size.py:14]: You should register the CI to the specific label like this:
https://github.com/sgl-project/sglang/blob/93109cc89be31809e45886377c55099983effa22/python/sglang/jit_kernel/tests/test_moe_lora_align_block_size.py#L14

**Review by yushengsu-thu — CHANGES_REQUESTED** [review]

File: python/sglang/jit_kernel/csrc/moe/moe_align_kernel.cu [hunk 2]
```diff
@@ -0,0 +1,572 @@
...
+    size_t numel,
+    bool pad_sorted_token_ids,
+    int32_t max_num_tokens_padded) {
+  if (threadIdx.x < fill_threads) {
+    if (pad_sorted_token_ids) {
+      for (int32_t it = threadIdx.x; it < max_num_tokens_padded; it += fill_threads) {
+        sorted_token_ids[it] = numel;
+      }
+    }
+    __syncthreads();
```
Comment (DarkSharpness) [inline python/sglang/jit_kernel/csrc/moe/moe_align_kernel.cu:277]: Why do we need this? Is this safe to call `__syncthreads` in different code paths? (It's not standard practice recommended in CUDA). Can we use other tilling strategy to avoid this (e.g. specificalize some blocks to do the alignment, instead of threads)

**Review by BBuf — APPROVED** [review]

File: python/sglang/jit_kernel/csrc/moe/moe_align_kernel.cu (hunk 2 above)
Comment (klshuster) [reply python/sglang/jit_kernel/csrc/moe/moe_align_kernel.cu:277]: Hi @DarkSharpness -- i failed to copy over the original comment above this block, but this is taken directly from the existing `moe_align_kernel` in `sgl-kernel`, which itself is adapted from a [vLLM PR](https://github.com/vllm-project/vllm/pull/29642/files#diff-5647b1413f4ae9aacba904eca8f8a8aee9079321eadff4c10101a2c6962dcc53R226). We can explore a different strategy but that would further distance this from the original implementation.

**Review by yushengsu-thu — APPROVED** [review]

**Review by Fridge003 — APPROVED** [review]
